File: app/src/main/java/com/ak47/www/koko_androiddemo/net/retrofit/RetrofitActivity.java
package com.ak47.www.koko_androiddemo.net.retrofit;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;

import com.ak47.www.koko_androiddemo.R;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_retrofit);

        simpleUse();


    }

    //简单实用
    public void simpleUse() {
        //第一步构建Retrofit
//        Retrofit retrofit = new Retrofit.Builder()
//                .baseUrl("https://api.github.com/")
//                .addConverterFactory(
//                        GsonConverterFactory.create()
//                )
//                .build();

        //第二步创建接口类
        // GitHubService service = retrofit.create(GitHubService.class);


        //封装避免重复创建对象
        GitHubService service = ServiceGenerator.createService(GitHubService.class);

        //第三步接口类调用抽象方法，返回Call类
        //https://api.github.com/users/smilekoko/repos
        Call<List<GitHubRepos>> call = service.listRepos("smilekoko");

        call.enqueue(new Callback<List<GitHubRepos>>() {
            @Override
            public void onResponse(Call<List<GitHubRepos>> call, Response<List<GitHubRepos>> response) {
                List<GitHubRepos> gitHubRepos = response.body();
                if (gitHubRepos != null) {
                    for (GitHubRepos repos : gitHubRepos) {
                        Log.e("koko", "Smilekoko的仓库id:" + repos.getId() + "///"
                                + "Smilekoko的仓库名字:" + repos.getName());
                    }

                }
            }

            @Override
            public void onFailure(Call<List<GitHubRepos>> call, Throwable t) {

            }
        });
    }
}

File: opensourcedemo/src/main/java/com/koko/opensourcedemo/retrofit/RetrofitActivity.kt
package com.koko.opensourcedemo.retrofit

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import com.koko.opensourcedemo.R
import retrofit2.Callback
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import com.koko.opensourcedemo.retrofit.bean.RepositoryJsonBean
import retrofit2.Call
import retrofit2.Response


class RetrofitActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_retrofit)


        val retrofit = Retrofit.Builder()
                .baseUrl("https://api.github.com/")
                .addConverterFactory(GsonConverterFactory.create())
                .build()


        getRepositoryListA(retrofit)

    }

    //异步请求
    private fun getRepositoryListA(retrofit: Retrofit) {
        val gitHubService = retrofit.create(GitHubService::class.java)
        val callRepos = gitHubService.jsonBeanListRepos("smilekoko")

//        https://api.github.com/users/smilekoko/repos
//        callRepos.enqueue(object : Callback<List<RepositoryBean>> {
//            override fun onResponse(call: Call<List<RepositoryBean>>, response: Response<List<RepositoryBean>>) {
//                val repositoryList = response.body()
//                if (repositoryList != null) {
//                    for (item in repositoryList)
//                        Log.e("koko", ":"+item.full_name)
//                }
//            }
//
//            override fun onFailure(call: Call<List<RepositoryBean>>, t: Throwable) {
//            }
//        })

        //使用Json来定义Bean
        callRepos.enqueue(object : Callback<List<RepositoryJsonBean>> {
            override fun onFailure(call: Call<List<RepositoryJsonBean>>, t: Throwable) {
                TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
            }

            override fun onResponse(call: Call<List<RepositoryJsonBean>>, response: Response<List<RepositoryJsonBean>>) {
                val repositoryList = response.body()
                if (repositoryList != null) {
                    for (item in repositoryList) {
                        Log.e("koko", item.fullName)
                    }
                }
            }
        }
        )
    }
}
